Fixed interaction where a Gemified card that gives a blue gem doesn't spend the correct resource amount when played

HumabHatterZed opened the pull request.

divisionbyz0rro commented:
@HumabHatterZed Regarding this fix: there's actually an edge case where this won't restore the vanilla game's behavior, and it has to do with how you wrote the check for updating the cost. If a card is granted the Gemify ability as part of a ResolveOnBoard trigger, you won't see that it has Gemify before you play it, and thus you will never reduce the cost. 

My recommendation would be to take the simpler approach: cache every cost before you play the card as you do now, then recalculate the costs *again* after playing the card, then charge the player the minimum of the two costs.

I've implemented this here as a reference: https://github.com/divisionbyz0rro/InscryptionAPI/blob/394a33d508d97c46bb4e9c428c851390bcee0e6d/InscryptionAPI/Costs/CardCostManager.cs#L472

HumabHatterZed commented:
> @HumabHatterZed Regarding this fix: there's actually an edge case where this won't restore the vanilla game's behavior, and it has to do with how you wrote the check for updating the cost. If a card is granted the Gemify ability as part of a ResolveOnBoard trigger, you won't see that it has Gemify before you play it, and thus you will never reduce the cost.
> 
> My recommendation would be to take the simpler approach: cache every cost before you play the card as you do now, then recalculate the costs _again_ after playing the card, then charge the player the minimum of the two costs.
> 
> I've implemented this here as a reference: https://github.com/divisionbyz0rro/InscryptionAPI/blob/394a33d508d97c46bb4e9c428c851390bcee0e6d/InscryptionAPI/Costs/CardCostManager.cs#L472

Added, ty <3
